File: validAnagram/validAnagram.ts
class Solution {
    sorting(s: string, t: string): boolean {
        if (s.length !== t.length) {
            return false;
        }
        return s.split('').sort().join('') === t.split('').sort().join('');
    }

    oneHash(s: string, t: string): boolean {
        if (s.length !== t.length) {
            return false;
        }
        const hash: Record<string, number> = {};
        for (let i = 0; i < s.length; i++) {
            hash[s[i]] = (hash[s[i]] || 0) + 1;
        }
        for (let i = 0; i < t.length; i++) {
            if (hash[t[i]] && hash[t[i]] > 0) {
                hash[t[i]] -= 1;
            } else {
                return false;
            }
        }
        return true;
    }

    twoHash(s: string, t: string): boolean {
        if (s.length !== t.length) {
            return false;
        }
        const hashS: Record<string, number> = {}, hashT: Record<string, number> = {};
        for (let i = 0; i < s.length; i++) {
            hashS[s[i]] = (hashS[s[i]] || 0) + 1;
            hashT[t[i]] = (hashT[t[i]] || 0) + 1;
        }
        for (const key in hashS) {
            if (hashS[key] !== hashT[key]) {
                return false;
            }
        }
        return true;
    }

    oneArray(s: string, t: string): boolean {
        if (s.length !== t.length) {
            return false;
        }
        const count: number[] = new Array(26).fill(0);
        for (let i = 0; i < s.length; i++) {
            count[s.charCodeAt(i) - 'a'.charCodeAt(0)] += 1;
            count[t.charCodeAt(i) - 'a'.charCodeAt(0)] -= 1;
        }
        for (const val of count) {
            if (val !== 0) {
                return false;
            }
        }
        return true;
    }
}
export default new Solution();

File: validAnagram/validAnagram.py
class ValidAnagram:
    def sorting(self, s, t):
        if len(s) != len(t):
            return False
        return sorted(s) == sorted(t)
    
    def oneHash(self, s, t):
        if len(s) != len(t):
            return False
        hash = {}
        for i in range(len(s)):
            if s[i] in hash:
                hash[s[i]] += 1
            else:
                hash[s[i]] = 1
        for i in range(len(t)):
            if t[i] in hash and hash[t[i]] > 0:
                hash[t[i]] -= 1
            else:
                return False
        return True

    def twoHash(self, s, t):
        if len(s) != len(t):
            return False
        hashS, hashT = {}, {}
        for i in range(len(s)):
            hashS[s[i]] = 1 + hashS.get(s[i], 0)
            hashT[t[i]] = 1 + hashT.get(t[i], 0)
        return hashS == hashT
    
    def oneArray(sefl, s, t):
        if len(s) != len(t):
            return False
        count = [0] * 26
        for i in range(len(s)):
            count[ord(s[i]) - ord('a')] += 1
            count[ord(t[i]) - ord('a')] -= 1
        for val in count:
            if val != 0:
                return False
        return True